lib/codegen/CreateTerm.cpp

```diff
@@ -0,0 +1,190 @@
+; map: %map
+; list: %list
+; set: %set
+; array: %list
+; integer: %mpz *
+; float: %mpfr *
+; string: %string *
+; bytes: %string *
+; string buffer: %string **
+; boolean: i1
+; machine integer of N bits: iN
+
+; We also define the following LLVM structure types:
+
+%string = type { %blockheader, [0 x i8] } ; 16-bit layout, 48-bit length, bytes
+%map = type { i64, i8 *, i8 * } ; im::hashmap::HashMap
+%set = type { i8 *, i8 *, i64 } ; im::hashset::HashSet
+%list = type { i64, i64, i8 *, i8 *, i8 *, i8 *, i8 * } ; im::vector::Vector
+%mpz = type { i32, i32, i64 * } ; mpz_t
+%mpfr = type { i64, i32, i64, i64 * } ; mpfr_t
+%blockheader = type { i64 } 
+%block = type { %blockheader, [0 x i64 *] } ; 16-bit layout, 8-bit length, 32-bit tag, children
+
+; The layout of a block uniquely identifies the categories of its children as
+; well as how to allocate/deallocate them and whether to follow their pointers
+; during gc. Roughly speaking, the following rules apply:
+
+; %string *: malloc/free, do not follow
+; iN: noop, do not follow
+; %map, %set, %list: noop/drop_in_place, follow
+; %block *: managed heap, follow
+; %mpz *: malloc/mpz_clear->free, do not follow
+; %mpfr *: malloc/mpfr_clear->free, do not follow
+; %string **: malloc->malloc/free->free, do not follow
+
+; We also automatically generate for each unique layout id a struct type
+; corresponding to the actual layout of that block. For example, if we have
+; the symbol symbol foo{Map{}, Int{}, Exp{}} : Exp{}, we would generate the type:
+
+; %layoutN = type { %blockheader, [0 x i64 *], %map, %mpz *, %block * }
+)LLVM";
+
+std::unique_ptr<llvm::Module> newModule(std::string name, llvm::LLVMContext &Context) {
+  llvm::SMDiagnostic Err;
+  return llvm::parseIR(*llvm::MemoryBuffer::getMemBuffer(LLVM_HEADER), Err, Context);
+}
+
+llvm::Type *getValueType(std::string sort, llvm::Module *Module) {
+  if (sort == "Map") {
+    return Module->getTypeByName("map");
+  } else if (sort == "List") {
+    return Module->getTypeByName("list");
+  } else if (sort == "Set") {
+    return Module->getTypeByName("set");
+  } else if (sort == "Array") {
+    return Module->getTypeByName("list");
+  } else if (sort == "Int") {
+    return llvm::PointerType::get(Module->getTypeByName("mpz"), 0);
+  } else if (sort == "Float") {
+    return llvm::PointerType::get(Module->getTypeByName("mpfr"), 0);
+  } else if (sort == "StringBuffer") {
+    return llvm::PointerType::get(llvm::PointerType::get(Module->getTypeByName("string"), 0), 0);
+  } else if (sort == "Bool") {
+    return llvm::IntegerType::get(Module->getContext(), 1);
+  } else if (sort == "MInt") {
+    assert(false && "not implemented yet: MInt");
+  } else {
+    return llvm::PointerType::get(Module->getTypeByName("block"), 0);
+  }
+}
+
+
+llvm::StructType *getBlockType(llvm::Module *Module, const KOREObjectSymbol *symbol, KOREObjectSymbolDeclaration *symbolDecl) {
+  llvm::StructType *BlockHeaderType = Module->getTypeByName("blockheader");
```

Comment:
Anywhere where we must absolutely use strings constants like here, they should be declared as globals to avoid mistyping.